src/main/java/org/ohdsi/webapi/shiro/filters/TeamProjectBasedAuthorizingFilter.java

```diff
@@ -0,0 +1,253 @@
+   * @param login: the user login
+   * @param request: the request passing through this filter
+   * @param response: response object that can be used to write error status and
+   * error message if needed.
+   * 
+   * @return: Returns true if the "team project" was found and passed validation
+   * and DB update to user roles did not fail.
+   * 
+   * @throws IOException
+   * @throws Exception
+   */
+  private static boolean extractAndValidateTeamProjectRoleAndUpdateUserIfNecessary(
+      TeamProjectBasedAuthorizingFilter self,
+      String login, ServletRequest request,
+      ServletResponse response)  throws IOException, Exception {
+
+    // synchronize on login to avoid race conditions (especially on DB updates) if
+    // this filter receives parallel requests for any reason:
+    synchronized (login.intern()) {
+      // check if a teamproject parameter is found in the request:
+      String teamProjectRole = self.extractTeamProjectFromRequestParameters(request);
+      Set<String> newUserRoles = new HashSet<String>();
+      Set<String> newDefaultRoles = new HashSet<String>(self.defaultRoles);
+
+      // if found, add teamproject as a role in the newUserRoles list:
+      if (teamProjectRole != null && !teamProjectRole.trim().isEmpty()) {
+        // double check if this role has really been granted to the user:
+        if (self.checkGen3Authorization(teamProjectRole, login) == false) {
+          String errorMessage = "User is not authorized to access this team project's data";
+          self.logger.error(errorMessage);
+          WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+            errorMessage);
+          return false;
+        }
+        // add teamproject role and related system role that
+        // enables read restrictions/permissions based read access configurations:
+        newDefaultRoles.add("read restricted Atlas Users"); // aka reserved system role 15 - TODO: since this is always added...maybe it should be one of the default roles instead...
+        newUserRoles.add(teamProjectRole);
+        self.authorizer.setCurrentTeamProjectRoleForCurrentUser(teamProjectRole, login);
+        self.authorizer.updateUser(login, newDefaultRoles, newUserRoles, true);
+        return true;
+      } else {
+        String errorMessage = "The teamproject is compulsory when on authorizationMode==teamproject configuration";
+        self.logger.error(errorMessage);
+        WebUtils.toHttp(response).sendError(HttpServletResponse.SC_FORBIDDEN,
+          errorMessage);
+        return false;
+      }
+    }
+  }
+
+
+  private boolean checkGen3Authorization(String teamProjectRole, String login) throws Exception {
+    logger.debug("Checking Gen3 Authorization for 'team project'={} and user={} using service={}", teamProjectRole, login, this.authorizationUrl);
+    RestTemplate restTemplate = new RestTemplate();
+    String arboristAuthorizationURL = this.authorizationUrl;
```

Comment:
The first paramater looks like it should be in the configuration file maybe

Reply:
which one? this.authorizationUrl? This does come from the pom or the properties file.

Reply:
Please ignore this comment. I think I was referring to the teamProjectRole but must have been confused since that is given to the WebAPI from Atlas.